betree/pmdk/src/lib.rs

```diff
@@ -0,0 +1,192 @@
+#![allow(non_upper_case_globals)]
+#![allow(non_camel_case_types)]
+#![allow(non_snake_case)]
+
+include!(concat!(env!("OUT_DIR"), "/bindings.rs"));
+
+use std::sync::{Arc, Mutex};
+use std::os::raw::c_void;
+use std::ptr::NonNull;
+
+#[derive(Debug)]
+//pub struct ptr_to_pmem(Arc<Mutex<*mut c_void>>);
+pub struct ptr_to_pmem(Arc<*mut c_void>);
+
+unsafe impl Send for ptr_to_pmem {}
+unsafe impl Sync for ptr_to_pmem {}
+
+pub mod libpmem {
+    use super::*;
+
+    pub fn pmem_file_create(filepath : &str, len: u64, mapped_len : &mut u64, is_pmem : &mut i32) -> Option<ptr_to_pmem> {
+        unsafe {
+            let mut ptr = pmem_map_file(filepath.as_ptr() as *const i8,
+                                     len,
+                                     (PMEM_FILE_CREATE|PMEM_FILE_EXCL) as i32,
+                                     0666,
+                                     mapped_len,
+                                     is_pmem);
+
+            match NonNull::new(ptr) {
+                //Some(value) => Some(ptr_to_pmem( Arc::new(Mutex::new(ptr)))),
+                Some(value) => Some(ptr_to_pmem( Arc::new(ptr))),
+                None => None
+            }
+        }
+    }
+
+    pub fn pmem_file_open(filepath: &str, mapped_len: &mut u64, is_pmem: &mut i32) -> Option<ptr_to_pmem> {
+        unsafe {
+            let mut ptr = pmem_map_file(filepath.as_ptr() as *const i8, 
+                                     0, // Opening an existing file requires no flag(s).
+                                     0, // No length as no flag is provided.
+                                     0666, 
+                                     mapped_len, 
+                                     is_pmem);
+
+            match NonNull::new(ptr) {
+                //Some(value) => Some(ptr_to_pmem( Arc::new(Mutex::new(ptr)))),
+                Some(value) => Some(ptr_to_pmem( Arc::new(ptr))),
+                None => None
+            }
+        }
+    }
+
+    pub fn pmem_file_read(begin_pos: &ptr_to_pmem, offset: usize, data: &mut [u8], len: u64) -> Result<(), std::io::Error>{
+        //match NonNull::new(*begin_pos.0.lock().unwrap()) {
+        match NonNull::new(*begin_pos.0) {
+            None => {
+                return Err(std::io::Error::new(std::io::ErrorKind::Other,
+                                       format!("File handle is missing for the PMEM file.")
+                                       ));
+            },
+            _ => ()
+        };
+
+        unsafe {
+            //let ptr = pmem_memcpy(data.as_ptr() as *mut c_void, begin_pos.0.lock().unwrap().add(offset), len, PMEM_F_MEM_NOFLUSH /*| PMEM_F_MEM_TEMPORAL*/);
```

Comment:
We can also remove the commented out code, since the code does not contain locking anymore.